### ISSUE TITLE
When using `isinstance` to infer a generic type in a block, use any covariant bounds on the type variable.

### ISSUE THREAD

**Issue by sixolet** [python/mypy#1998]
I have a class hierarchy like the following:

```python
class Apple(object): ...

T_co = TypeVar('T', covariant=True)
A_co = TypeVar('A_co', bound=Apple, covariant=True)

class Action(Generic[T_co]): ...

class AppleAction(Generic[A_co], Action[A_co]): ...
```

Later in my code, I have a block like so:

```python
if isinstance(action, AppleAction):
   reveal_type(action)
```

And I get `AppleAction[Any]`.  Because `A_co` is covariant, and bounded by `Apple`, even if `A_co` is bound to a subtype of `Apple`, the resulting object is going to be a valid instance of `AppleAction[Apple]`.  I'd like to have `action` be `AppleAction[Apple]` in this case.

**Comment by rwbarton** [comment]
I agree. This really has more to do with the covariance than the bound, since every type variable has an upper bound, which is usually just `object`. So if there was no upper bound on `A_co`, I'd expect to get an `AppleAction[object]` out here.

Actually I think the fact that `isinstance` produces `Any` types like it does is always wrong; but it's only easy to fix in this case when the class is covariant (or when the class is contravariant).

**Comment by ilevkivskyi** [comment]
This is still happening, indeed it looks like we just need to fix this for covariant type variables.
